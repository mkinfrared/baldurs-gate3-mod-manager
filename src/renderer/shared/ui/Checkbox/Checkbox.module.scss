@import "../../styles/vars";

.Checkbox {
  --background-color: transparent;
  --border-color: var(--primary-4);
  --border-radius: 0.2rem;
  --border-width: 0.5rem;

  @media (prefers-color-scheme: light) {
    --border-color: var(--primary-7);
  }

  // TODO ask Natasha is 2rem is fine for checkmark block
  // needs to be adjusted because the text is 2rem tall
  --checkbox-width: 2rem;
  --icon-color: transparent;
  --stroke-color: var(--primary-5);
  --stroke-dashoffset: 25;
  --text-color: var(--primary-10);
  --timing-duration: 250ms;
  --transition-time: 250ms;

  align-items: center;
  cursor: pointer;
  display: flex;
  width: fit-content;

  &:has(input:checked) {
    --icon-color: var(--primary-4);

    @media (prefers-color-scheme: light) {
      --icon-color: var(--primary-7);
    }

    .tick {
      path {
        stroke-dashoffset: 0;
        transition:
          var(--timing-duration) stroke-dashoffset linear,
          var(--timing-duration) fill linear var(--timing-duration);
      }
    }
  }

  // https://stackoverflow.com/questions/70375526/why-is-read-only-css-pseudo-class-being-applied-on-this-checkbox
  &:has(input:disabled, input[readonly]) {
    --stroke-color: var(--gray-88);
    --text-color: var(--gray-88);

    filter: grayscale(0.8);
    pointer-events: none;
  }

  &.error:has(.input:is(.error)) {
    --border-color: var(--danger-4);
    --stroke-color: var(--danger-4);
    --text-color: var(--danger-4);

    @media (prefers-color-scheme: light) {
      --border-color: var(--danger-7);
      --stroke-color: var(--danger-7);
      --text-color: var(--danger-7);
    }

    &:has(input:checked) {
      --icon-color: var(--danger-4);

      @media (prefers-color-scheme: light) {
        --icon-color: var(--danger-7);
      }
    }
  }

  .container {
    align-items: center;
    border-radius: 0.4rem;
    display: flex;
  }

  .input {
    height: 0;
    margin: 0;
    opacity: 0;
    position: absolute;
    width: 0;
    z-index: -1;
  }

  .input:focus-visible ~ .container {
    @include focus-visible;
  }

  .text {
    color: var(--text-color);
    margin-left: 1rem;
  }

  .box {
    align-items: baseline;
    aspect-ratio: 1/1;
    background: transparent;
    border: 1px solid var(--border-color);
    border-radius: var(--border-radius);
    display: flex;
    height: var(--checkbox-width);
    justify-content: center;
    padding: 0.3rem;
    position: relative;
    width: var(--checkbox-width);
  }

  .tick {
    fill: var(--icon-color);
    height: 100%;
    opacity: 1;
    overflow: visible;
    width: 100%;

    path {
      fill: var(--icon-color);
      stroke: var(--stroke-color);
      stroke-dasharray: 25;
      stroke-dashoffset: var(--stroke-dashoffset);
      stroke-opacity: 1;
      stroke-width: 0.2;
      transition:
        var(--timing-duration) fill linear,
        var(--timing-duration) stroke-dashoffset linear var(--timing-duration);
    }
  }
}
