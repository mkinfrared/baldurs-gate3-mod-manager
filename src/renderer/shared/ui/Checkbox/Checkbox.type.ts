export type CheckboxProps = {
  /**
   * a string that will be applied as a css class to parent element
   */
  className?: string;
  checked?: boolean;
  defaultChecked?: boolean;
  disabled?: boolean;
  error?: boolean;
  name?: string;
  inputRef?: React.Ref<HTMLInputElement | null | undefined>;
  /**
   * label text to render next to checkbox
   */
  label?: string;
  onChange?: React.ChangeEventHandler<HTMLInputElement>;
  readOnly?: boolean;
  value?: string;
};
