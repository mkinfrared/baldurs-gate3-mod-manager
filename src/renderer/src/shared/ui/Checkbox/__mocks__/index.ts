export { default } from "./Checkbox.mock";
