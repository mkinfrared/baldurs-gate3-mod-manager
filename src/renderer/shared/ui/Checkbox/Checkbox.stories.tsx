/* eslint-disable react-hooks/rules-of-hooks */
import { action } from "@storybook/addon-actions";
import { Meta, StoryObj } from "@storybook/react";
import { ChangeEvent, useState } from "react";

import { Checkbox } from "./Checkbox";
import css from "./Story.module.scss";

type Story = StoryObj<typeof meta>;

const meta = {
  title: "UI/Checkbox",
  component: Checkbox,
  tags: ["autodocs"],
} satisfies Meta<typeof Checkbox>;

const Uncontrolled: Story = {
  args: {
    onChange: action("change"),
    label: "marklar",
  },
  render: (args) => (
    <div className={css.Story}>
      <Checkbox {...args} />
    </div>
  ),
};

const Controlled: Story = {
  render: () => {
    const [checked, setChecked] = useState(false);

    const handleChange = (event: ChangeEvent<HTMLInputElement>) => {
      const { target } = event;

      setChecked(target.checked);
    };

    return (
      <div className={css.Story}>
        <Checkbox name="example" onChange={handleChange} checked={checked} />
      </div>
    );
  },
};

export { Uncontrolled, Controlled };

export default meta;
