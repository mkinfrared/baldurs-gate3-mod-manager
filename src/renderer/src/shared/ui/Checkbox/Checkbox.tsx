import Tick from "@renderer/shared/assets/icons/tick.svg";
import { classNames, mergeRefs } from "@renderer/shared/lib/helpers";
import Text from "@renderer/shared/ui/Text";

import css from "./Checkbox.module.scss";
import { CheckboxProps } from "./Checkbox.type";

/**
 * Renders an html checkbox
 */
const Checkbox = ({
  className,
  checked,
  defaultChecked,
  disabled = false,
  error = false,
  inputRef,
  label,
  name,
  onChange,
  readOnly = false,
  value,
}: CheckboxProps) => {
  const mergedRefs = mergeRefs(inputRef);

  return (
    <label
      className={classNames(css.Checkbox, error && css.error, className)}
      data-testid="Checkbox"
    >
      <input
        className={classNames(css.input, error && css.error)}
        checked={checked}
        defaultChecked={defaultChecked}
        disabled={disabled}
        name={name}
        onChange={onChange}
        ref={mergedRefs}
        readOnly={readOnly}
        type="checkbox"
        value={value}
      />
      <div className={css.container}>
        <div className={css.box}>
          <Tick className={css.tick} />
        </div>
        {label && <Text className={css.text}>{label}</Text>}
      </div>
    </label>
  );
};

export { Checkbox };

export default Checkbox;
