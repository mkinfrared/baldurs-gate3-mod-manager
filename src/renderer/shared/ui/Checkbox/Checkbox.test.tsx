import { render } from "@testing-library/react";

import { Checkbox } from "./Checkbox";

describe("<Checkbox />", () => {
  const Component = <Checkbox />;

  it("should be defined", () => {
    expect(Checkbox).toBeDefined();
  });

  it("should match the snapshot", () => {
    const { container } = render(Component);

    expect(container).toMatchSnapshot();
  });
});
