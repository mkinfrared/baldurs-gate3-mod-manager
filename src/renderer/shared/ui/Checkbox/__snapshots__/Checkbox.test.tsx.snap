// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox /> should match the snapshot 1`] = `
<div>
  <label
    class="Checkbox"
    data-testid="Checkbox"
  >
    <input
      class="input"
      type="checkbox"
      value=""
    />
    <div
      class="container"
    >
      <div
        class="box"
      >
        <svg
          class="tick"
        >
          tick.svg
        </svg>
      </div>
    </div>
  </label>
</div>
`;
